node/src/components/storage.rs

```diff
@@ -1,859 +1,578 @@
+                let mut tx = self.env.begin_rw_txn()?;
+                let outcome = tx.put_value(self.deploy_db, deploy.id(), &deploy, false)?;
+                tx.commit()?;
+                responder.respond(outcome).ignore()
             }
-            Event::Request(StorageRequest::GetDeploys {
+            StorageRequest::GetDeploys {
                 deploy_hashes,
                 responder,
-            }) => self.get_deploys(deploy_hashes, responder),
-            Event::Request(StorageRequest::GetDeployHeaders {
+            } => responder
+                .respond(self.get_deploys(&mut self.env.begin_ro_txn()?, deploy_hashes.as_slice())?)
+                .ignore(),
+            StorageRequest::GetDeployHeaders {
                 deploy_hashes,
                 responder,
-            }) => self.get_deploy_headers(deploy_hashes, responder),
-            Event::Request(StorageRequest::PutExecutionResults {
+            } => responder
+                .respond(
+                    // TODO: Similary to getting block headers, requires optimized function.
+                    self.get_deploys(&mut self.env.begin_ro_txn()?, deploy_hashes.as_slice())?
+                        .into_iter()
+                        .map(|opt| opt.map(|deploy| deploy.header().clone()))
+                        .collect(),
+                )
+                .ignore(),
+            StorageRequest::PutExecutionResults {
                 block_hash,
                 execution_results,
                 responder,
-            }) => self.put_execution_results(block_hash, execution_results, responder),
-            Event::Request(StorageRequest::GetDeployAndMetadata {
+            } => {
+                let mut tx = self.env.begin_rw_txn()?;
+
+                for (deploy_hash, execution_result) in execution_results {
+                    let mut metadata = self
+                        .get_deploy_metadata(&mut tx, &deploy_hash)?
+                        .unwrap_or_default();
+
+                    // If we have a previous execution result, we enforce that it is the same.
+                    if let Some(prev) = metadata.execution_results.get(&block_hash) {
+                        if prev != &execution_result {
+                            return Err(Error::DuplicateExecutionResult {
+                                deploy_hash,
+                                block_hash,
+                            });
+                        }
+
+                        // We can now skip adding, as the result is the same.
+                        continue;
+                    }
+
+                    // Update metadata and write back to db.
+                    metadata
+                        .execution_results
+                        .insert(block_hash, execution_result);
+                    let was_written =
+                        tx.put_value(self.deploy_metadata_db, &deploy_hash, &metadata, true)?;
+                    assert!(was_written);
```

Comment:
This will _always_ be false, unless I gravely misunderstood what I am doing. The "no panic" policy does not include "no panics on actual implementation bugs".

Reply:
You can fault me for false advertising though :)

Reply:
I believe it _is_ false advertising - the docs explicitly mention potential bugs are catered for via fatal errors :)  I won't belabour the point though.  If you're comfortable with it, I can let it slide.